source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.cc

```diff
@@ -0,0 +1,109 @@
+#include "resource_provider.h"
+
+#include <string>
+
+#include "source/common/common/logger.h"
+#include "source/common/config/utility.h"
+
+#include "resource_detector.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+
+namespace {
+bool isEmptyResource(const Resource& resource) { return resource.attributes.empty(); }
+
+Resource createInitialResource(std::string service_name) {
+  Resource resource{};
+
+  // Creates initial resource with the static service.name attribute.
+  if (service_name.empty()) {
+    service_name = std::string{kDefaultServiceName};
+  }
+  resource.attributes[std::string(kServiceNameKey.data(), kServiceNameKey.size())] = service_name;
+  return resource;
+}
+
+/**
+ * @brief Calculates the new schema url when merging two resources.
+ * This function implements the algorightm as defined in the OpenTelemetry Resource SDK
+ * specification. @see
+ * https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/resource/sdk.md#merge
+ *
+ * @param old_schema_url The old resource's schema URL.
+ * @param updating_schema_url The updating resource's schema URL.
+ * @return std::string The calculated schema URL.
+ */
+std::string mergeSchemaUrl(const std::string& old_schema_url,
+                           const std::string& updating_schema_url) {
+  if (old_schema_url.empty()) {
+    return updating_schema_url;
+  }
+  if (updating_schema_url.empty()) {
+    return old_schema_url;
+  }
+  if (old_schema_url == updating_schema_url) {
+    return old_schema_url;
+  }
+  // The OTel spec leaves this case (when both have value but are different) unspecified.
+  ENVOY_LOG_MISC(info, "Resource schemaUrl conflict. Fall-back to old schema url: {}",
+                 old_schema_url);
+  return old_schema_url;
+}
+
+/**
+ * @brief Updates an old resource with a new one. This function implements
+ * the Merge operation defined in the OpenTelemetry Resource SDK specification.
+ * @see
+ * https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/resource/sdk.md#merge
+ *
+ * @param old_resource The old resource.
+ * @param updating_resource The new resource.
+ */
+void mergeResource(Resource& old_resource, const Resource& updating_resource) {
+  if (isEmptyResource(updating_resource)) {
```

Comment:
Actually, the [spec ](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/resource/sdk.md#merge)somewhat mentions the case where both resources have a `schemaUrl` set but different:

>Else this is a merging error (this is the case when the Schema URL of the old and updating resources are not empty and are different). The resulting resource is undefined, and its contents are implementation-specific.

And the collector impl https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/resourcedetection.go#L160 returns the old schema. Essentially I think this will result in the same behavior I have. If `updating_resource` has a schema URL but no resources, in the collector the current resource will have still the current schemaUrl and not from the updating one.

Reply:
> old_resource had an empty-schema, the merged resource will receive the updating_resource schema.
Yes, if `old_resource` has an empty schema and the `updating_resource` has a non-empty schema, the updating should be used. BUT in your case the `updating_resource` has also no attributes and to me, updating the old resource with the schema from the updating in this case is wrong, since the updating resource has no attributes. The schema describes the definition/conventions of the attributes it carries, and if it's empty then I'd consider it a wrong situation. I can try to bring this to the OTel spec to seek further clarification but that may take a while.

Reply:
@adisuissa for now, I changed the behavior in Envoy to follow the spec exactly - schema URL is merged regardless of a resource having attributes or not [512e692](https://github.com/envoyproxy/envoy/pull/29547/commits/512e692dd7eda9779b1086333e6c828432209fd2).

In the meantime I opened an issue in the OTel spec https://github.com/open-telemetry/opentelemetry-specification/issues/3710. If something there changes I will update here as a follow up.

Reply:
ACK, thanks! Please add a TODO followup in this code, so one can trace back this comment if needed in the future.

Reply:
Done!